### PR TITLE
u-boot: 2023.07.02 -> 2023.10

### PR DESCRIPTION
**Pull request by jmbaur**
## Description of changes

Among other changes, this release moves more boards to use the new environment format where a simple text file is used under ./board rather than using CFG_EXTRA_ENV_SETTINGS. This includes the default environment for the raspberry pi, hence the change of patch.

Something I haven't been able to track down is the need for removing the DTC make flag. U-boot has a version of this (along with libfdt) in their tree that builds just fine, so I propose we use that instead.

I've tested these changes building various boards defined in nixpkgs (cross compiling from x86_64-linux) with good success. I'm not sure of the expectation that every u-boot board defined in nixpkgs build successfully, (as I saw one with an incorrect `filesToInstall`), so I have not tested all boards. Please test if these changes break your board!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jmbaur** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.ubootRaspberryPi4_64bit

**Comment by jmbaur** [comment]
@ofborg build ubootTools

**Review by samueldr — CHANGES_REQUESTED** [review]
(No changes requested)

All devices should be tried to be built (pkgsCross on their arch) before merging.

It is fine to assume `built == valid`, given the track record of U-Boot.

**Comment by jmbaur** [comment]
I quickly built all immediately build-able (with a pkgsCross package set available) u-boot derivations and gathered the results here: https://gist.github.com/jmbaur/466df3f4200914f906c05994d3f38674. All of them that fail also fail on current master. Luckily the failure is the same, so the fix should be easy. I'm happy to try and track down the SCP binary for these boards and include that in this PR, I'm assuming this means they will need to be marked unfree. I'm also happy keeping them broken and fixing them in a separate PR, just let me know. I also fixed cross-compilation of ubootTools.

**Comment by lopsided98** [comment]
This PR fixes the SCP build failures: https://github.com/NixOS/nixpkgs/pull/252111

**Comment by lopsided98** [comment]
These are the attributes for the builds with question marks:
```
pkgsCross.raspberryPi.ubootRaspberryPi
pkgsCross.raspberryPi.ubootRaspberryPiZero
pkgsCross.sheevaplug.ubootSheevaplug
pkgsCross.sheevaplug.ubootGuruplug
```

**Comment by jmbaur** [comment]
Thanks @lopsided98! The rest of them build successfully (gist updated), and that PR you linked would cover the rest that are failing.

**Comment by lopsided98** [comment]
Successfully booted:
* `ubootRaspberryPi2`
* `ubootRockPro64`: Initially failed with `Error enabling VMMC supply : -114`, fixed after applying https://patchwork.ozlabs.org/project/uboot/list/?series=364753&state=*
* `ubootRock64v2` (from #191538): works with above patch applied (didn't test without, but it is probably necessary)
* `ubootOdroidC2`

Failed:
* `ubootRaspberryPi4_64bit`: can't debug right now, but fails to boot
* `ubootOdroidXU3`: not tested, but known bugs have not been fixed, see: https://github.com/NixOS/nixpkgs/pull/227947#issuecomment-1668654983

**Comment by samueldr** [comment]
When looking for Tow-Boot, I've observed that all Raspberry Pi boards were moved to `bootstd`. So depending on the exact setup, it might be related to that move.

**Comment by lopsided98** [comment]
I thought I had a saved environment on this RPi4, which might have broken things with bootstd, but this wasn't the problem.

The real problem is that, by default, U-Boot has `boot_targets=mmc usb pxe dhcp`. This appears to make it try mmc 1 first, which is an unconnected interface on the RPi4. This obviously fails, but rather than trying mmc 0, it just moves on to USB and PXE. Explicitly specifying `boot_targets=mmc0` makes it boot successfully. Interestingly, https://github.com/u-boot/u-boot/commit/c771e5b8c2a186fb072b6c6f571d4a3cc86efba9 says:

> We don't need to specify the mmc devices individually, since they are
> used in order from 0 to 2, and standard boot uses that order anyway.

This appears to be incorrect, and the fix is probably to set `boot_targets=mmc0 mmc1 mmc2 usb pxe dhcp`

**Comment by lopsided98** [comment]
Another odd RPi4 bug: the 2 second boot delay takes closer to 10 seconds

**Comment by jmbaur** [comment]
I added a patch to address this, let me know if you think this is appropriate or not. Unfortunately I do not have an rpi to try and reproduce this, but I'll check out the behavior on qemu with bootstd.

**Comment by lopsided98** [comment]
I submitted the patch upstream; we'll see what they think: https://patchwork.ozlabs.org/project/uboot/patch/20231007155535.1699567-1-benwolsieffer@gmail.com/

File: pkgs/misc/uboot/default.nix [hunk 1]
```diff
@@ -48,7 +47,8 @@ let
     src = if src == null then defaultSrc else src;
 
     patches = [
-      ./0001-configs-rpi-allow-for-bigger-kernels.patch
+      ./0001-rpi.env-allow-for-bigger-kernels.patch
+      ./0002-rpi.env-try-to-boot-from-mmc0-first.patch
```
Comment (samueldr) [inline pkgs/misc/uboot/default.nix:51]: To keep track of this new situation.

```suggestion
      # https://patchwork.ozlabs.org/project/uboot/patch/20231007155535.1699567-1-benwolsieffer@gmail.com/
      ./0002-rpi.env-try-to-boot-from-mmc0-first.patch
```

**Comment by lopsided98** [comment]
Upstream suggested trying this patch, which also fixes the RPi4 issue: https://patchwork.ozlabs.org/project/uboot/patch/20230923205017.1754340-1-sjg@chromium.org/

**Comment by jmbaur** [comment]
I included the patch here: https://patchwork.ozlabs.org/project/uboot/patch/20230923205017.1754340-1-sjg@chromium.org/raw

**Comment by sorki** [comment]
I missed this one and created https://github.com/NixOS/nixpkgs/pull/261775 which does bunch of other things as well. 

Except for `DTC=dtc` removal the first patch is pretty much identical
- I've added another patch fixing QEMU test
- We can still use external `dtc` but it has to be a full path otherwise it fails horribly.

@jmbaur do you mind if I pick the 3 other patches to my branch and we continue there?

**Comment by jmbaur** [comment]
@sorki that sounds good to me!

**Comment by jmbaur** [comment]
closed in favor of #261775

**Comment by sorki** [comment]
Much appreciated @jmbaur!

**Pull request closed without merging**